File: delta-shop-app-ui/src/app/core/services/review.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { PageResponse } from './product.service';
import { environment } from '../../../environments/environment';

export interface Review {
  id: string;
  productId: string;
  productName: string;
  userId: string;
  userName: string;
  userAvatar: string;
  rating: number;
  status: 'PENDING' | 'APPROVED' | 'REJECTED';
  title: string;
  body: string;
  images: string[];
  verifiedPurchase: boolean;
  helpfulCount: number;
  adminReply: string;
  adminReplyAt: string;
  createdAt: string;
  votedHelpful?: boolean;
}

export interface ReviewEligibility {
  orderItemId: string;
  productId: string | null;
  canReview: boolean;
  reviewed: boolean;
  reason: string | null;
  reviewId: string | null;
  reviewStatus: 'PENDING' | 'APPROVED' | 'REJECTED' | null;
}

export interface ReviewStats {
  averageRating: number;
  totalReviews: number;
  fiveStarCount: number;
  fourStarCount: number;
  threeStarCount: number;
  twoStarCount: number;
  oneStarCount: number;
  verifiedPurchaseCount: number;
  withImagesCount: number;
}

@Injectable({
  providedIn: 'root'
})
export class ReviewService {
  private apiUrl = `${environment.apiUrl}/reviews`;

  constructor(private http: HttpClient) { }

  getProductReviews(productId: string, page: number = 0, size: number = 10, sortBy: string = 'createdAt'): Observable<PageResponse<Review>> {
    const params = new HttpParams()
      .set('page', String(page))
      .set('size', String(size))
      .set('sortBy', sortBy);
    return this.http.get<PageResponse<Review>>(`${this.apiUrl}/products/${productId}`, { params });
  }

  getProductReviewStats(productId: string): Observable<ReviewStats> {
    return this.http.get<ReviewStats>(`${this.apiUrl}/products/${productId}/stats`);
  }

  createReview(payload: { productId: string; orderItemId: string; rating: number; title?: string; body?: string; images?: string[] }): Observable<Review> {
    return this.http.post<Review>(this.apiUrl, payload);
  }

  getReviewEligibility(orderItemId: string): Observable<ReviewEligibility> {
    return this.http.get<ReviewEligibility>(`${this.apiUrl}/order-items/${orderItemId}/eligibility`);
  }

  getProductReviewEligibility(productId: string): Observable<ReviewEligibility> {
    return this.http.get<ReviewEligibility>(`${this.apiUrl}/eligibility/products/${productId}`);
  }

  updateReview(reviewId: string, payload: { rating?: number; title?: string; body?: string }): Observable<Review> {
    return this.http.put<Review>(`${this.apiUrl}/${reviewId}`, payload);
  }

  deleteReview(reviewId: string): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${reviewId}`);
  }

  voteHelpful(reviewId: string, helpful: boolean): Observable<void> {
    const params = new HttpParams().set('helpful', String(helpful));
    return this.http.post<void>(`${this.apiUrl}/${reviewId}/helpful`, {}, { params });
  }

  getMyReviews(page: number = 0, size: number = 10): Observable<PageResponse<Review>> {
    const params = new HttpParams()
      .set('page', String(page))
      .set('size', String(size));
    return this.http.get<PageResponse<Review>>(`${this.apiUrl}/me`, { params });
  }

  getPendingReviews(page: number = 0, size: number = 20): Observable<PageResponse<Review>> {
    const params = new HttpParams()
      .set('page', String(page))
      .set('size', String(size));
    return this.http.get<PageResponse<Review>>(`${this.apiUrl}/admin/pending`, { params });
  }

  moderateReview(reviewId: string, payload: { status: string; adminReply?: string }): Observable<Review> {
    return this.http.post<Review>(`${this.apiUrl}/admin/${reviewId}/moderate`, payload);
  }

  batchModerateReviews(reviewIds: string[], status: string): Observable<void> {
    return this.http.post<void>(`${this.apiUrl}/admin/bulk-moderate`, { reviewIds, status });
  }
}

File: src/main/java/com/hoangdinh/delta_shop_app/service/ReviewService.java
package com.hoangdinh.delta_shop_app.service;

import com.hoangdinh.delta_shop_app.dto.request.review.ReviewCreateRequest;
import com.hoangdinh.delta_shop_app.dto.request.review.ReviewUpdateRequest;
import com.hoangdinh.delta_shop_app.dto.request.review.ReviewModerationRequest;
import com.hoangdinh.delta_shop_app.dto.response.PageResponse;
import com.hoangdinh.delta_shop_app.dto.response.review.ReviewResponse;
import com.hoangdinh.delta_shop_app.dto.response.review.ReviewStatsResponse;
import com.hoangdinh.delta_shop_app.dto.response.review.ReviewEligibilityResponse;

import java.util.List;
import java.util.UUID;

public interface ReviewService {

    // User operations
    ReviewResponse createReview(UUID userId, ReviewCreateRequest request);
    ReviewResponse updateReview(UUID reviewId, UUID userId, ReviewUpdateRequest request);
    void deleteReview(UUID reviewId, UUID userId);
    void voteHelpful(UUID reviewId, UUID userId, boolean helpful);

    // Query operations
    PageResponse<ReviewResponse> getProductReviews(UUID productId, UUID userId, int page, int size, String sortBy);
    PageResponse<ReviewResponse> getUserReviews(UUID userId, int page, int size);
    ReviewResponse getReviewById(UUID reviewId);
    ReviewStatsResponse getProductReviewStats(UUID productId);
    ReviewEligibilityResponse getReviewEligibility(UUID userId, UUID orderItemId);
    ReviewEligibilityResponse getProductReviewEligibility(UUID userId, UUID productId);

    // Admin operations
    PageResponse<ReviewResponse> getPendingReviews(int page, int size);
    ReviewResponse moderateReview(UUID reviewId, ReviewModerationRequest request, UUID adminId);
    void batchModerateReviews(List<UUID> reviewIds, String status, UUID adminId);

    // Helper methods
    boolean canUserReviewProduct(UUID userId, UUID productId);
    void checkAndUpdateProductRating(UUID productId);
}
